cumulus/parachains/runtimes/testing/penpal/src/pallets/transact_helper.rs

```diff
@@ -0,0 +1,117 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_core::H160;
+	use sp_std::{vec, vec::Vec};
+	use xcm::prelude::*;
+	use xcm_executor::traits::XcmAssetTransfers;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type XcmRouter: SendXcm;
+		type XcmExecutor: ExecuteXcm<Self::RuntimeCall> + XcmAssetTransfers;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// ExportMessage message was sent
+		SentExportMessage { message_id: XcmHash, sender_cost: Assets, message: Xcm<()> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InvalidMsg,
+		FeesNotMet,
+		SendFailure,
+	}
+
+	#[derive(Clone, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo)]
+	pub struct TransactInfo {
+		pub target: H160,
+		pub call: Vec<u8>,
+		pub gas_limit: u64,
+		pub fee: u128,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		[u8; 32]: From<<T as frame_system::Config>::AccountId>,
+	{
+		#[pallet::call_index(0)]
+		#[pallet::weight(Weight::from_parts(100_000_000, 0))]
+		pub fn transact_to_ethereum(
+			origin: OriginFor<T>,
+			target: H160,
+			call: Vec<u8>,
+			fee: u128,
+			gas_limit: u64,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let dest = Location {
+				parents: 2,
+				interior: Junctions::from([GlobalConsensus(Ethereum { chain_id: 11155111 })]),
+			};
+			let transact = TransactInfo { target, call, gas_limit, fee };
+
+			let inner_message = Xcm(vec![
+				Transact {
+					origin_kind: OriginKind::SovereignAccount,
+					require_weight_at_most: Weight::default(),
+					call: transact.encode().into(),
+				},
+				SetTopic([0; 32]),
+			]);
+
+			let (ticket, price) =
+				validate_send::<T::XcmRouter>(dest.clone(), inner_message.clone())
+					.map_err(|_| Error::<T>::InvalidMsg)?;
+			ensure!(price.len() > 0, Error::<T>::FeesNotMet);
+
+			let mut fees: Assets = (Parent, fee).into();
+			fees.push(price.get(0).unwrap().clone());
+
+			let origin = Location::from(AccountId32 { network: None, id: who.into() });
+
+			T::XcmExecutor::charge_fees(origin, fees.clone().into())
```

Comment:
Charge fees sends it to the treasury account. Is this the behavior we want for the delivery fees?

Reply:
It's for local execution on Penpal. On BH For delivery we charge from sovereign of penpal for the local execution cost in DOT. 
```
2024-04-22T04:24:47.085187Z TRACE xcm::process: origin: Some(Location { parents: 1, interior: X1([Parachain(2000)]) }), total_surplus/refunded: Weight { ref_time: 0, proof_size: 0 }/Weight { ref_time: 0, proof_size: 0 }, error_handler_weight: Weight { ref_time: 0, proof_size: 0 }
2024-04-22T04:24:47.085200Z TRACE xcm::process_instruction: === WithdrawAsset(Assets([Asset { id: AssetId(Location { parents: 1, interior: Here }), fun: Fungible(4000000000) }]))
2024-04-22T04:24:47.085209Z TRACE xcm::ensure_can_subsume_assets: worst_case_holding_len: 1, holding_limit: 64
2024-04-22T04:24:47.085250Z TRACE xcm::fungible_adapter: withdraw_asset what: Asset { id: AssetId(Location { parents: 1, interior: Here }), fun: Fungible(4000000000) }, who: Location { parents: 1, interior: X1([Parachain(2000)]) }
2024-04-22T04:24:47.085368Z TRACE xcm::process_instruction: === BuyExecution { fees: Asset { id: AssetId(Location { parents: 1, interior: Here }), fun: Fungible(4000000000) }, weight_limit: Limited(Weight { ref_time: 729581493, proof_size: 13316 }) }
2024-04-22T04:24:47.085385Z TRACE xcm::weight: UsingComponents::buy_weight weight: Weight { ref_time: 729581493, proof_size: 13316 }, payment: AssetsInHolding { fungible: {AssetId(Location { parents: 1, interior: Here }): 4000000000}, non_fungible: {} }, context: XcmContext { origin: Some(Location { parents: 1, interior: X1([Parachain(2000)]) }), message_id: [106, 51, 55, 199, 155, 107, 230, 240, 194, 5, 253, 187, 40, 174, 94, 95, 214, 45, 219, 36, 8, 138, 134, 194, 34, 165, 69, 28, 207, 132, 148, 30], topic: None }
```